README
Apassword_oracle

Short explanation:

An "oracle" is an external service you can ask questions of.  A
"password oracle" is one that answers questions about password
quality.

This password oracle tracks whether or not a password as been assigned
"recently" where "recent" is tunable and by default "not within the
last 65,536 password allocations." 

It uses a novel data-structure call a "deprecating sketch".  This is a
type of counting bloom-filter and somewhat similar but different than
the "counting-sketch" that Schechter, Herley and Mitzenmacher propose
in http://research.microsoft.com/pubs/132859/popularityISeverything.pdf

This algorithum differed from Schechter et al in how the counters are
used.  Their sketch tracks total membership and occationally increases
a global threshold.  Mine maintains a ring buffer that tracks old
assignments and decrements counters when the password expires from the
ring buffer.  Schechter et al allows "credit" to be built up so a
password can be used many times in succession after a long period of
disuse.  My approach does not allow this.

What makes these class of "password popularity" checkers so neat is if
an adversary downloads a copy of the server's database they will not
be able to extract any useful passwords from it.

Instructions on how to use this server can be found in the docstring
of password_oracle.py

Long explanation: 

Way back in the olden days of computing password security meant
/etc/shadow files, salts, and picking passwords that would be hard for
a determined cracker to guess given a small cluster of faster
computers.

Times have changed.  No longer is your adversary a student picking away
at the internal security of their school's timeshare Unix machine.
The new adversary is in the cloud.

Modern online "cloud" services have a number of tools at their
disposal to prevent automated password guessing attacks on their
networks.  Captcha, per IP address throttle limits and per account
throttle limits.

Some sites (like Google) use captcha to block automated password login
attempts.  In the long run this won't work, computer vision is
improving at a scary fast pace, human vision slogs along to the whims
of evolutionary pressure over thousands and millions of years.  Pretty
soon we'll be scratching our heads and asking our personal AI's to
tell us what the captcha says.  Scratch that.

IP based throttling doesn't work - there is simply no go way to tell
if a particular IP is a household on a DSL, a large companies office
or an entire nation.  IP address exhaustion doesn't help this matter.
And bad-guys have no qualms using bot-nets to spread their nefarious work
across lots and lots of IP addresses.

The only thing that works is "user-name" based throttling.  For the
most part 1990's style password brute forcing can be prevented by
throttling the rate at which the doorknob to any users account can be
turned.

Put yourself in the shoes of a cracker.  You have unlimited IP
addresses.  The only real limit you have is how many passwords you can
try for one account.  How do you maximize your haul.

Pick common passwords.  In the Rock You password corpus over 1% of all
users selected "123456" as their password - the most efficient way to
steal accounts is to just try "123456" on each account and move on if
it doesn't work ... if facebook is the same that should net you a cool
half a million or so accounts.  

So, how do you defend against this?  

You could restrict how often passwords are reused.  Twitter disallows
their 390 most common passwords.  But there are few problems with
that.  First, it requires you have your passwords stored someplace in
the clear so you can figure out what your most common passwords are.
Companies make headlines doing that.

The second problem is users are really predictable in how they modify
their passwords to address password restrictions.  Guess what number
50% of all users that put a number in the middle of their passwords
choose - 4.  And if you look at the text that comes before and after
the "4" its obvious "4" is a stand-in for "for."  Tell a user "you
can't use 123456" and they are likely to use "1234567" instead.  Or
write in "l33tsp3ak."  Simple password restrictions simply shift the
peak around.  You also don't want ti make password selection too
difficult; if the user has to think too hard they are likely to go
away and use another service.  Ideally you want to dynamically measure
the number of times a password occurs and deny the user when the
password they are selecting is "too common", but 

So how do we choose between these two evils; living without the
ability to know which passwords are common and having clear text
passwords?

Turns out we can have our cake and eat it to with a "bloom-filter." 

A bloom-filter is a probabilistic data structure that tracks
approximate set membership; it trades off accuracy and the
possibility it will accidentally claim an item is in a set that it
isn't in exchange for a dramatically reduced storage space.
Bloom-filters have a nice sweet spot where they throw away enough
information that if a bad guy got hold of the filter they would not be
able to reconstruct the set membership (i.e. the clear-text passwords)
but at the same time we have a low enough false positive rate that the
filter is still useful.

A traditional bloom-filter is a write-once structure; once you add
something you can't take it out.  Its okay if more than one user uses
the same password, we just don't want any one password to be used too
often.  This algorithm changes that somewhat, instead of using bits,
we store integers and maintain a deque that tracks when it is time to
remove items from the bloom-filter again.

This won't be the first algorithm to approach this problem.  There is
a fair amount of prior work dating back to the 90's about bloom-filters
and passwords.  Microsoft research recently published a paper
proposing a different approach called a counting sketch. 

The difference between the counting sketch and the deprecating sketch
is how never before used passwords are treated on a long running
system.  Suppose you have a sketch configured to not allow any
password to be used for more than one in every thousand accounts, and
your service as grown to the point where it has a million users.
Suppose a new user created 10 accounts in a short span of time and
tried to assign them all the same albeit never seen before password.
A counting sketch would allow this.  A deprecating sketch would not.
 LocalWords:  docstring


NEWS
Feb 26th, 2011: Added support for the transmission of hash values

The password-oracle was designed to be resistant to database
inspection attacks by storing approximate records of passwords in a
bloomfilter, but the original API required you pass clear text
passwords to the password-oracle.  That's great if you are running
this only your only server or a private rack with a private intranet on
a switch off of your own, but it downgrades the security I worked so
hard to acquire by using a bloomfilter in the first place.

password-oracle now accepts hash values instead of clear text
passwords.  Here's how it works.

When your application starts up, ask it for the "divisor" you should
use to generate a hash value.  This number will depend on your bloom
filter size configuration parameters, and is the number of slots you
have raised to the number of samples per password.

# curl --get 127.0.0.1:8000/hash_range.json
 123456 


Because this value is items ** per_item, you should consider setting
per_item to 1.  This will reduce the accuracy of your bloomfilter
slight, but make it much less likely that an adversary will be able to
extract passwords from your communication with the server.

Now to use the server, your application, whether in java script or in
your app server, will need to somehow generate a hash value, store it
as a number and mod it by this value.  

So, if in python you might write:

>>> int(hashlib.md5('secret').hexdigest(), 16) % 123456 
61225 

Use this value as a proxy for your password.  If you use the default
settings, but set --per-item to 1, this value will be 65536 and
generally won't be reversible.

Now use the "available" and "add" commands as you did before, except
with the parameter hash=61225 instead of password=secret

Note that entropy.json won't work because it can't reverse the hash and figure
out your original password (there are gazillions of good matches).
all won't work because it calls entropy.

Also keep in mind that unless you are absolutely sure you are using
the same hash functions password_oracle does, you cannot meaningfully
mix calls with password and hash; switching, or switching your hash
algorithm will basically result in gibberish and random behavior
until your current passwords deprecate.  The system won't be
unusable, you'll still have the same false positive rate, it just
won't catch your common passwords the first time they are used, its
sort of like stating with a fresh bloomfilter except you have a false
positive rate from the get-go instead of building up gradually as
items are added.


utils_test.py
#!/usr/bin/env python2.6 

import unittest
from utils import * 

class PrefixTest(unittest.TestCase):
    def test_prefixed(self):
        self.assertEquals(prefixed("abc","a"),
                          "bc")

    def test_not_prefixed(self):
        self.assertEquals(prefixed("abc", "b"),
                          False)

class SegmentationTest(unittest.TestCase):
    def test_segment(self):
        expected = [((None, None), "s"),
                    ((None, "s"), "t"),
                    (("s", "t"), "r"),
                    (("t", "r"), "i"),
                    (("r", "i"), "n"),
                    (("i", "n"), "g"),
                    (("n", "g"), None),
                    (("g", None), None)]
        actual = list(segment("string"))
        self.assertEqual(expected, actual)

    def test_segment_with_non_default_length(self):
        expected = [((None,), "s"),
                    (("s",), None)]
        actual = list(segment("s", 2))
        self.assertEqual(expected, actual)

if __name__ == "__main__":
    unittest.main()



language_model_test.py
#!/usr/bin/env python2.6 

from language_model import *
import unittest
import StringIO


class LanguageModelTest(unittest.TestCase):
    def setUp(self):
        self.language_model = compile(StringIO.StringIO("aaa\naab\nabb\naaa"))
    def test_compile(self):
        expected = {('a', 'a'): {'a': 2, 'b': 1, None: 2}, 
                    ('b', 'b'): {None: 1}, 
                    (None, None): {'a': 4}, 
                    ('a', None): {None: 2}, 
                    ('a', 'b'): {'b': 1, None: 1}, 
                    ('b', None): {None: 2}, 
                    (None, 'a'): {'a': 3, 'b': 1}}

        self.assertEqual(expected, self.language_model)

    def test_entropy(self):
        self.assertAlmostEquals(self.language_model.bits("aaa"), 2.0)
        self.assertAlmostEquals(self.language_model.bits("aab"), 3.32, 2)

    def test_pickling(self):
        self.assertEqual(cPickle.loads(cPickle.dumps(self.language_model)),
                         self.language_model)
    
    def test_construction(self):
        self.assertEqual(LanguageModel(self.language_model),
                         self.language_model)
        
if __name__ == "__main__":
    unittest.main()
    


password_oracle_test.py
#!/usr/bin/env python2.6 

import unittest
import deprecating_sketch
import language_model 
import StringIO
from password_oracle import * 

PREFIX = "/prefix/"

class FakeServer:
    def __init__(self, sketch, language_model):
        self.sketch = sketch
        self.language_model = language_model 
        

class PasswordOracleRequestHandlerCrashDummy(PasswordOracleRequestHandler):
    def __init__(self, sketch, language_model):
        self.server = FakeServer(sketch, language_model) 
        self.headers_ended = False 
        self.wfile = StringIO.StringIO()
        self.test_password = None
        self.test_hash = None

    def path_prefix(self):
        return PREFIX 

    def send_response(self, *value):
        self.response_code = value 

    def end_headers(self):
        self.headers_ended = True 

    def get_post_password(self):
        # Not bothering to test cgi.FieldStorage.  Pretty sure it works. 
        return self.test_password

    def get_post_hash(self):
        return self.test_hash


class PasswordOracleRequestHandlerTest(unittest.TestCase):
    def setUp(self):
        self.handler = PasswordOracleRequestHandlerCrashDummy(
            deprecating_sketch.DeprecatingSketch(slots=1000, items=2, per_item=1),
            language_model.compile(StringIO.StringIO("aaa\naab\nabb\naaa")))

    def test_get_fails_on_bad_prefix(self):
        self.handler.path = '/abc/'
        self.handler.do_GET()
        self.assertEquals(self.handler.response_code[0], 404)
    
    def test_post_fails_on_bad_prefix(self):
        self.handler.path = '/abc/'
        self.handler.test_password='secret'
        self.handler.do_POST()
        self.assertEquals(self.handler.response_code[0], 404)

    def test_has_entropy(self):
        self.handler.path = PREFIX + "entropy.json?password=aaa"
        self.handler.do_GET()
        self.assertEquals(self.handler.response_code[0], 200)
        self.assertAlmostEqual(json.loads(self.handler.wfile.getvalue()), 2.0)

    def test_missing_entropy_skipped(self):

        self.handler.path = PREFIX + "entropy.json?password=aaa"
        self.handler.data = dict(password="aaa")
        self.handler.server.language_model.clear()
        self.handler.do_GET()
        self.assertEquals(self.handler.response_code[0], 503)

    def test_get_all(self):
        self.handler.path = PREFIX + "all.json?password=aaa"
        self.handler.do_GET()
        actual = json.loads(self.handler.wfile.getvalue())
        self.assertAlmostEqual(actual['entropy'], 2.00)
        self.assertEqual(actual['available'], True)

    def test_get_bits_required(self):
        self.handler.path = PREFIX + "hash_range.json?password"
        self.handler.do_GET()
        actual = json.loads(self.handler.wfile.getvalue())
        self.assertEqual(actual, 1000)

class PasswordOracleRequestHandlerComplexTest(unittest.TestCase):
    def setUp(self):
        self.sketch = deprecating_sketch.DeprecatingSketch(slots=100, items=2, per_item=1)

        self.handler = PasswordOracleRequestHandlerCrashDummy(
            self.sketch, 
            language_model.compile(StringIO.StringIO("aaa\naab\nabb\naaa")))
        
    def test_post_writes_to_sketch(self):
        sketch = self.sketch
        handler = self.handler 

        handler.path = PREFIX + "available.json?password=secret"
        handler.data = dict(password="secret")
        handler.do_GET()
        self.assertEquals(handler.response_code[0], 200)
        self.assertTrue(json.loads(handler.wfile.getvalue()))

        handler.path = PREFIX + "add"
        handler.test_password='secret'
        handler.do_POST()
        self.assertEquals(handler.response_code[0], 201)

        handler = PasswordOracleRequestHandlerCrashDummy(
            sketch, 
            language_model.compile(StringIO.StringIO("aaa\naab\nabb\naaa")))

        handler.path = PREFIX + "available.json?password=secret"
        handler.do_GET()
        self.assertFalse(json.loads(handler.wfile.getvalue()))

    def test_hashmode_writes_to_sketch(self):
        sketch = self.sketch
        handler = self.handler 

        handler.path = PREFIX + "available.json?hash=1"
        handler.data = dict(hash="1")
        self.assertEquals(self.handler.get_hash(), 1 )
        handler.do_GET()
        self.assertEquals(handler.response_code[0], 200)
        self.assertTrue(json.loads(handler.wfile.getvalue()))

        handler.path = PREFIX + "add" 
        handler.test_hash = 1 
        handler.do_POST()
        self.assertEquals(handler.response_code[0], 201)

        handler = PasswordOracleRequestHandlerCrashDummy(
            sketch, 
            language_model.compile(StringIO.StringIO("aaa\naab\nabb\naaa")))

        handler.path = PREFIX + "available.json?hash=1"
        handler.do_GET()
        self.assertFalse(json.loads(handler.wfile.getvalue()))

if __name__ == "__main__":
    unittest.main()


utils.py
#!/usr/bin/env python2.6 


class Identity:
    """A mixin class that provides custom hashing and comparison

    Hashing and comparisons are performed on the results of a new method
    __identity__.   So, instead of writing:

    def __hash__(self):
      return hash(dosomething(self))

    def __cmp__(self, other):
      return cmp(dosomething(self), dosomething(other))

    You would simply define:

    def __identity__(self):
      return dosomething(self)
    """

    def __hash__(self):
        hash(self.__identity__())

    def __cmp__(self, other):
        return cmp(self.__identity__(), other.__identity__())

def segment(s, n=3):
    """Segments a string s into substrings of length n

    Args:
      s: string to segment
      n: segment length (defaults to 3)

    Yields:
      Individual segments of s as described below   
    
    Example:
    
    segment("ab") yields:
      [None, None, "a"]
      [None, "a", "b"]
      ["a", "b", None]
      ["b", None]
      """
    padded_string = [None] * (n-1) + list(s) + [None] * (n-1)
    
    for offset in range(len(padded_string) - n + 1):
        yield (tuple(padded_string[offset : offset + n - 1]), padded_string[offset + n - 1])


def all_but_the_last(generator):
    """Yields all but the last item in a sequence.

    This is just like list[generator][:-1] except it doesn't require
    the entire sequnce be precomputed.
    """ 
    second_to_last = generator.next()
    while True:
        last = generator.next()
        yield second_to_last
        second_to_last = last 


def prefixed(s, prefix):
    """Returns the non-prefixed form of s if it starts with prefix.

    Args:
      s: String to examine
      prefix: Prefix to test for

    Returns:
      False is s does not start with prefix, s with the prefix removed otherwise
    """
    return s.startswith(prefix) and s[len(prefix):]


password_oracle.py
#!/usr/bin/env python2.6 

"""password_oracle

A password quality oracle implemented as a HTTP server.  Given a
password it will answer whether the password has been used recently
and how "common" the password is compared to a list of common words.

For documentation of command line parameters:

./pasword_oracle.py --help 

Tutorial:

1. Compile a word list as follows.  A short list that is quick to
compile and download is provided, but if you are patient I recommend
you download and use the Rockyou password corpus.  Also if your list
is long be patient, on a "typical" machine (MacBook Air 1.86ghz) the
compiler will process about 1/2 million lines per minute Also note
that repeated words increase the weight of that word.

# grep -v '#' password.lst |  ./language_model.py | \
  gzip -9 > language_model.pickle.gz 

2. Start your server.  Note that changing bloom-filter parameters will
erase your recent password history.  The server will by default store
its history data in "bloom_filter.pickle", but by default no language
model is loaded.

BTW, by default the bloom-filter tables are setup for a 65,536 element
history.  This means that no password can occur more than 1 out of
ever 65,536 times.  --items indicates how many items are actually
stored at once, while I change it for the demo, changing it without
changing the other parameters will lead to security problems -- with
only three items and the other settings as they are it is actually
possible to figure out what those three passwords are (slots should
be a small single digit multiple of items * per_item, and per_items
should be a small single digit number.)  But we change items to make
the demo easier.


# ./password_oracle.py --language_model=language_model.pickle  --items=3

Now in another window lets ask the oracle if 123456 has recently been
used as a password:

# curl --get -d password=123456 127.0.0.1:8000/available.json
true 

It is available, so it hasn't been used recently.  Lets ask what the
oracle thinks about the quality of the password (this demo assumes you
loaded the RockYou password list, not the silly demo provided in
passwords.lst.)

# curl --get -d password=123456 127.0.0.1:8000/entropy.json
6.9068905956085187

6.9 bits of entropy.  Not to good.  Hey, try this, ask about 12345

# curl --get -d password=12345 127.0.0.1:8000/entropy.json; echo 
8.2288186904958813

Neat, huh?  123456 is so common, so much more common than passwords
than end in 45 that adding that extra 6 actually reduces the entropy
of your password.  Longer passwords are not automatically better than
shorter passwords.  Lets get off this tangent write to the
bloom-filter.

Because 123456 is available, I've decided to use it:

# curl  -d password=123456 127.0.0.1:8000/add

There is nothing here preventing two people from requesting the same
password at the same time.  See the README file for a discussion of
why this isn't a problem.

Now lets see if this password is available.e

# curl --get -d password=123456 127.0.0.1:8000/available.json
false

One of the features of a deprecating sketch is it throttles the use of
passwords, it isn't a forever ban.  Because --items=3, "123456" will
be forgotten after three more passwords.

# curl  -d password=1234 127.0.0.1:8000/add
# curl --get -d password=123456 127.0.0.1:8000/available.json
false
# curl  -d password=12345 127.0.0.1:8000/add
# curl --get -d password=123456 127.0.0.1:8000/available.json
false
# curl  -d password=12345 127.0.0.1:8000/add
# curl --get -d password=123456 127.0.0.1:8000/available.json
true


Caveats

  Changing the deprecating sketch's thresholds will zero out its
  underlying bloom filter.

  We're passing passwords in the clear with this API.  I hope this
  server isn't facing the public Internet and talking to a java script
  client in the user's browser.

  When adding passwords this program doesn't care if you are adding
  the same or different passwords; old entries are flushed after
  however many adds are specified in --items.
"""

import BaseHTTPServer
import cPickle
import cgi
import deprecating_sketch
import gflags 
import gzip
import json
import language_model
import signal 
import sys 
import urlparse 
import utils

from language_model import LanguageModel, Histogram

GFLAGS = gflags.FLAGS 

gflags.DEFINE_string('path', '/', 'URL prefix.')
gflags.DEFINE_string('host', '', 'Host addr to listen to')
gflags.DEFINE_integer('port', 8000, 'Port addr to listen to')
gflags.DEFINE_string('language_model', None, 'Language model to load')
gflags.DEFINE_string('bloom_filter', 'bloom_filter.pickle', 'Bloomfilter to load')


HTTP_UNAVAILABLE = 503
HTTP_OK = 200 
HTTP_CREATED = 201 
HTTP_NOT_FOUND = 404
HTTP_BAD_FORMAT = 415 

ERR_INTERRUPTED = 4 


class PasswordRequired:
    "Raised to note that a particular handler requires a password"

class BadPrefix:
    "Raised if the URI prefix isn't correct (wrong branch.)"

class PasswordOracleRequestHandler(BaseHTTPServer.BaseHTTPRequestHandler):
    """PasswordORacleRequestHander
    
    An http interface to:
    
    * Check membership in the deprecating hash (actually returns
      !member, true if the password is available for use, false
      otherwise.)

      GET PREFIX/available.json?password=123456 -> bool
      GET PREFIX/available.json?hash=37 -> bool

    * Add temporary membership to the deprecating hash

      POST PREFIX/add password=123456
      POST PREFIX/add hash=36

    * Check the entropy of a password against a fixed precompiled
      language model (more bits is better, for a large web service
      your usres should have ~15-20 bits minimum)

      GET PREFIX/entropy.json?password=123456 -> float 

    * Get both membership and entropy in one convenience call.  Should
      be a little faster than calling available and entropy
      sequentially
      GET PREFIX/all.json&password=123 -> dict(entropy=float, available=bool)

    * Find this database's required divisor for hash values
      GET PREFIX/get_hash.json -> 65536 

    Only .json is supported right now.
    """

    def path_prefix(self):
        """Get the path prefix from GFLAGS.path.  
        
        This is basically a dependency injection point for testing.
        """
        return GFLAGS.path

    def get_command(self):
        "Returns the current command (i.e. all, entropy, etc etc.)"
        scheme, netloc, path,  params, query, fragment = urlparse.urlparse(self.path)
        if utils.prefixed(path, self.path_prefix()):
            return path[len(self.path_prefix()):]
        raise BadPrefix()

    def get_password(self):
        "get_password returns the password for a GET request."
        scheme, netloc, path,  params, query, fragment = urlparse.urlparse(self.path)
        data = cgi.parse_qs(query)
        password = data.get('password')
        if password:
            return password[0]

    def get_hash(self):
        "get_password returns the hash for a GET request."
        scheme, netloc, path,  params, query, fragment = urlparse.urlparse(self.path)
        data = cgi.parse_qs(query)
        hash = data.get('hash')
        if hash:
            return int(hash[0])

    def get_post_password(self):
        "get_post_password returns the password for a POST requiest."
        form = cgi.FieldStorage(
            fp=self.rfile, 
            headers=self.headers,
            environ={'REQUEST_METHOD':'POST',
                     'CONTENT_TYPE':self.headers['Content-Type'],
                     })
        return form['password'].value

    def compute_entropy(self):
        """compute_entropy
        
        Returns the entropy in a password, or None if no language
        model was loaded.
        """
        if not self.get_password():
            raise PasswordRequired()

        if self.server.language_model:
            return self.server.language_model.bits(self.get_password())

    def compute_available(self):
        """compute_available
        
        Compute the "availability" of a password based on its
        membership in the deprecating sketch.
        
        Returns:
          False if the password was added "recently", True otherwise.
        """
        password = self.get_password() or self.get_hash()
        if not password:
            raise PasswordRequired()
        return password not in self.server.sketch

    def compute_all(self):
        """compute_all
        
        Computes both the entrpy and "availability" of a password
        
        Returns: 
          dict(entropy = float,
               available = true/false)
        """
        return dict(entropy=self.compute_entropy(),
                    available=self.compute_available())

    def compute_hash_range(self):
        return self.server.sketch.hash_range

    def password_required(self, password):
        "Raises PasswordRequired if not password"
        if self.get_password() or self.get_hash():
            return
        raise PasswordRequired()

    def do_GET(self):
        "Handle GET requests"
        try:
            function, format = self.get_command().split('.', 2)
        except BadPrefix:
            return self.send_response(HTTP_NOT_FOUND, 'Wrong prefix')
        
        function = {'entropy':self.compute_entropy,
                    'available':self.compute_available,
                    'hash_range':self.compute_hash_range,
                    'all':self.compute_all}.get(function)

        format = {'json': json.dumps}.get(format)
        
        if not function:
            return self.send_response(HTTP_NOT_FOUND, 'Unknown function')
        if not format:
            return self.send_response(HTTP_BAD_FORMAT, 'Unknown format')
        try:
            data = function()
        except PasswordRequired:
            self.send_response(HTTP_NOT_FOUND, 'No password provided')
            return 

        if data is None:
            return self.send_response(HTTP_UNAVAILABLE)
        self.send_response(HTTP_OK)
        self.end_headers()
        
        self.wfile.write(format(data))


    def do_POST(self):
        "Handle POST requests"
        try:
            if self.get_command() != 'add':
                return self.send_response(HTTP_NOT_FOUND, 'Unknown command')
            password = self.get_post_password()
            hash = self.get_post_hash()
            if not (password or hash):
                return self.send_response(HTTP_NOT_FOUND, 'Missing password')

            self.send_response(HTTP_CREATED)
            self.end_headers()
            self.server.sketch.add(password or hash)
        except BadPrefix:
            return self.send_response(HTTP_NOT_FOUND, 'Bad Prefix')
        

class PasswordOracleServer(BaseHTTPServer.HTTPServer):
    """PasswordOracleServer
    
    An HTTPServer that preloads a deprecating sketch and optionally a static language model. 
    
    The deprecating hash is saved on SIGTERM or Ctrl-C
    """
    
    @staticmethod 
    def load(pathname, default_class, open=open):
        """Load a pickle based configuration file.
        
        Args:
          pathname: The pathname of the file
          default_class: A class to instantiate if the config file
            cannot be loaded.
          open: An optional parameter indicating the function used to
            open the config file.  Useful for specifying a compressor
            such as gzip.open
        """
        if pathname:
            try:
                return cPickle.load(open(pathname))
            except Exception, ex:
                pass
        return default_class()

    @classmethod
    def sketch_factory(cls, sketch_path):
        """Load the current deprecating sketch."""
        return cls.load(sketch_path, deprecating_sketch.DeprecatingSketch)
    
    @classmethod
    def language_model_factory(cls, language_model_path):
        """Load the language_model.  Decompress it with gzip."""
        return cls.load(language_model_path, language_model.LanguageModel, open=gzip.open)

    def save(self, *_):
        """Save the current deprecating sketch."""
        cPickle.dump(self.sketch, open(self.sketch_path, "w+"))

    def __init__(self, sketch_path, language_model_path=None, *args, **kwargs):
        """Create a new instance of the PasswordOracleServer.
        
        Args:
          sketch_path: Path to the deprecating sketch.  A new sketch
            will be created on save if none currently exists.

          language_model_path: The path to the language model.
          Defaults to None in which case calls to "entropy" will
          return a 503
        """

        BaseHTTPServer.HTTPServer.__init__(self, *args, **kwargs)
        self.sketch = self.sketch_factory(sketch_path)
        self.language_model = self.language_model_factory(language_model_path)

        self.sketch_path = sketch_path

    def run_forever(self):
        """Run this service for ever.

        Catches and saves the deprecating sketch state on
        KeyboardInterrupt and signal.SIGTERM.
        """
        import select 
        signal.signal(signal.SIGTERM, self.save)
        try:
            while True:
                try:
                    self.handle_request()
                except select.error, err:
                    # Side effect of signal catching ... 
                    if (err[0], err[1]) != (ERR_INTERRUPTED, 'Interrupted system call'):
                        raise err
        except KeyboardInterrupt:
            self.save()

def main(argv):
    try:
        argv = GFLAGS(argv)  # parse flags
    except gflags.FlagsError, e:
        print '%s\nUsage: %s ARGS\n%s' % (e, sys.argv[0], GFLAGS)
        sys.exit(1)
    server_address = (GFLAGS.host, GFLAGS.port)
    PasswordOracleServer(GFLAGS.language_model, 
                         GFLAGS.bloom_filter, 
                         server_address, 
                         PasswordOracleRequestHandler).run_forever()


if __name__ == "__main__":
    main(sys.argv)
